postgres/postgres.go

```diff
@@ -0,0 +1,207 @@
+	addS3Url             *sql.Stmt
+	addEvent             *sql.Stmt
+	addDimension         *sql.Stmt
+	getDimensions        *sql.Stmt
+	addNodeID            *sql.Stmt
+	createPublishMessage *sql.Stmt
+}
+
+func prepare(sql string, db *sql.DB) *sql.Stmt {
+	statement, err := db.Prepare(sql)
+	if err != nil {
+		panic(err)
+	}
+	return statement
+}
+
+// NewDatastore - Create a postgres datastore. This is used to store and find information about jobs and instances.
+func NewDatastore(db *sql.DB) (Datastore, error) {
+	addJob := prepare("INSERT INTO Jobs(job) VALUES($1) RETURNING jobId", db)
+	updateState := prepare("UPDATE Jobs SET job = jsonb_set(job, '{state}', TO_JSONB($1::TEXT)) WHERE jobId = $2 RETURNING jobId", db)
+	addInstance := prepare("INSERT INTO Instances(jobId, instance) VALUES($1, $2) RETURNING instanceId", db)
+	findInstance := prepare("SELECT instance FROM Instances WHERE instanceId = $1", db)
+	addS3Url := prepare("UPDATE Jobs SET job = jsonb_set(job, '{files}', (SELECT (job->'files')  || TO_JSONB(json_build_object('alaisName',$1::TEXT,'url',$2::TEXT)) FROM Jobs WHERE jobId = $3), true) WHERE jobId = $3 RETURNING jobId", db)
+	addEvent := prepare("UPDATE Instances SET instance = jsonb_set(instance, '{events}', (SELECT (instance->'events')  || TO_JSONB(json_build_object('type', $1::TEXT, 'time', $2::TEXT, 'message', $3::TEXT, 'messageOffset', $4::TEXT)) FROM Instances WHERE instanceid = $5), true) WHERE instanceid = $5 RETURNING instanceId", db)
+	addDimension := prepare("INSERT INTO Dimensions(instanceId, nodeName, value) VALUES($1, $2, $3)", db)
+	getDimensions := prepare("SELECT nodeName, value, nodeId FROM Dimensions WHERE instanceId = $1", db)
+	addNodeID := prepare("UPDATE Dimensions SET nodeId = $1 WHERE instanceId = $2 AND nodeName = $3 RETURNING instanceId", db)
+	createPublishMessage := prepare("SELECT job->'recipe', job->'s3Files', STRING_AGG(instanceId::TEXT, ', ') FROM Jobs INNER JOIN  Instances ON (Jobs.jobId = Instances.jobId) WHERE jobs.jobId = $1 GROUP BY jobs.job", db)
+	return Datastore{db: db, addJob: addJob, updateState: updateState, addInstance: addInstance,
+		findInstance: findInstance, addS3Url: addS3Url, addEvent: addEvent, addDimension: addDimension,
+		getDimensions: getDimensions, addNodeID: addNodeID, createPublishMessage: createPublishMessage}, nil
+}
+
+// AddJob - Add a job to be stored in postgres.
+func (ds Datastore) AddJob(newjob *models.NewJob) (models.JobInstance, error) {
+	job := models.Job{Recipe: newjob.Recipe,
+		UploadedFileS: []models.UploadedFile{}, State: "Created"}
+	bytes, error := json.Marshal(job)
+	if error != nil {
+		return models.JobInstance{}, error
+	}
+	row := ds.addJob.QueryRow(bytes)
+	var jobID sql.NullString
+	rowError := row.Scan(&jobID)
+	if rowError != nil {
+		return models.JobInstance{}, rowError
+	}
+	instanceIDs := []string{}
+	for i := 0; i < newjob.NumberOfInstances; i++ {
+		id, instanceIDErr := ds.AddInstance(jobID.String)
+		if instanceIDErr != nil {
```

Comment:
Consider creating a transaction, so if creating an instance fails, we can rollback the entire job creation? Maybe not for initial implementation, but something to consider longer term.